Content.Server/GameObjects/Components/Radiation/RadiationPulseComponent.cs

```diff
@@ -0,0 +1,215 @@
+            serializer.DataField(this, x => x.Energy, "energy", 20.0f);
+            serializer.DataField(this, x => x.Range, "range", 5.0f);
+            serializer.DataField(this, x => x.Decay, "decay", true);
+            serializer.DataField(this, x => x.MinPulseLifespan, "minPulseLifespan", 3.0f);
+            serializer.DataField(this, x => x.MaxPulseLifespan, "maxPulseLifespan", 6.0f);
+            serializer.DataField(this, x => x.MinCooldownAct, "minCooldownAct", 0.2f);
+            serializer.DataField(this, x => x.MaxCooldownAct, "maxCooldownAct", 0.6f);
+        }
+
+        public void Update(float frameTime)
+        {
+            _timeLeftUntilRadiation += frameTime;
+
+            if (!Decay || Owner.Deleted)
+            {
+                return;
+            }
+
+            if (_duration <= 0.0f)
+            {
+                Owner.Delete();
+            }
+
+            _duration -= frameTime;
+        }
+
+        /// <summary>
+        /// Checks if the radiation pulse can radiate entities around it.
+        /// If the period between pulses was bigger than the cooldown,
+        /// the ratio will adjust the amount of radiation done.
+        /// </summary>
+        public bool OnCooldown(out float ratio)
+        {
+            if (_timeLeftUntilRadiation >= Cooldown)
+            {
+                ratio = _timeLeftUntilRadiation / Cooldown;
+                _timeLeftUntilRadiation = 0.0f;
+                return false;
+            }
+
+            ratio = default;
+            return true;
+        }
+
+        public abstract bool CanRadiate(IEntity entity);
+    }
+
+    [RegisterComponent]
+    [ComponentReference(typeof(RadiationPulseComponent))]
+    public sealed class RadiationPulseAnomaly : RadiationPulseComponent
+    {
+        public override uint? NetID => ContentNetIDs.RADIATION_PULSE;
+        public override string Name => "RadiationPulseAnomaly";
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            Appearance?.SetData(RadiationPulseVisual.State, RadiationPulseVisuals.Visible);
+        }
+
+        public override bool CanRadiate(IEntity entity)
+        {
+            return this.InRangeUnOccluded(entity, Range);
+        }
+
+        public override ComponentState GetComponentState()
+        {
+            return new RadiationPulseAnomalyState(Range, StartTime, EndTime);
+        }
+    }
+
+    [RegisterComponent]
+    [ComponentReference(typeof(RadiationPulseComponent))]
+    public sealed class RadiationPulseSingularity : RadiationPulseComponent
+    {
+        public override string Name => "RadiationPulseSingularity";
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            Appearance?.SetData(RadiationPulseVisual.State, RadiationPulseVisuals.None);
+        }
+
+        public override bool CanRadiate(IEntity entity)
+        {
+            return true;
+        }
+    }
```

Comment:
I still think you could wrap this into RadiationPulseComponent, same with Anomaly. I sussed it out with @PaulRitter for a 2nd opinion as well https://discord.com/channels/310555209753690112/770682801607278632/806456013468205056

I don't think the appearance data even changes does it? The component seems redundant given it's only set once in initialize when you could just have default values for it in yaml.

Reply:
imo i wouldn't even define additional behaviour for the singularity, i'd just make the singularity use the default behaviour. its already op as heck, no use in making it even more damaging